src/DotNetWorker.Core/Context/DefaultBindingCache.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    internal class DefaultBindingCache<T> : IBindingCache<T>
+    {
+        private readonly ConcurrentDictionary<string, T> _cache = new();
+
+        public bool TryGetValue(string key, out T? value)
+        {
+            EnsureKeyNotNull(key);
+
+            if (_cache.TryGetValue(key, out var conversionResult))
+            {
+                value = conversionResult;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        public bool TryAdd(string key, T value)
```

Comment:
We could also add a GetOrAdd which takes the key and a value factory func. But that can come as an enhancement later. This type is internal.